services-custom/iam-policy-builder/src/main/java/software/amazon/awssdk/policybuilder/iam/internal/DefaultIamPolicyReader.java

```diff
@@ -0,0 +1,212 @@
+        JsonNode principalsNode = statementObject.get(name);
+
+        if (principalsNode == null) {
+            return null;
+        }
+
+        if (principalsNode.isString() && principalsNode.asString().equals(IamPrincipal.ALL.id())) {
+            return singletonList(IamPrincipal.ALL);
+        }
+
+        if (principalsNode.isObject()) {
+            List<IamPrincipal> result = new ArrayList<>();
+            principalsNode.asObject().forEach((id, value) -> {
+                result.add(IamPrincipal.create(id, expectString(value, name + " entry value")));
+            });
+            return result;
+        }
+
+        throw new IllegalArgumentException(name + " was not \"" + IamPrincipal.ALL.id() + "\" or an object");
+    }
+
+    private List<IamCondition> readConditions(JsonNode conditionNode) {
+        if (conditionNode == null) {
+            return null;
+        }
+
+        Map<String, JsonNode> conditionObject = expectObject(conditionNode, "Condition");
+
+        List<IamCondition> result = new ArrayList<>();
+
+        conditionObject.forEach((operator, keyValueNode) -> {
+            Map<String, JsonNode> keyValueObject = expectObject(keyValueNode, "Condition key");
+            keyValueObject.forEach((key, value) -> {
+                if (value.isString()) {
+                    result.add(IamCondition.create(operator, key, value.asString()));
+                } else if (value.isArray()) {
+                    List<String> values =
+                        value.asArray()
+                             .stream()
+                             .map(valueNode -> expectString(valueNode, "Condition values entry"))
+                             .collect(toList());
+                    result.addAll(IamCondition.createAll(operator, key, values));
+                }
+            });
+
+        });
+
+        return result;
+    }
+
+    private List<String> readStringArray(Map<String, JsonNode> statementObject, String nodeKey) {
+        JsonNode node = statementObject.get(nodeKey);
+
+        if (node == null) {
+            return null;
+        }
+
+        if (node.isString()) {
+            return singletonList(node.asString());
+        }
+
+        if (node.isArray()) {
+            return node.asArray()
+                       .stream()
+                       .map(n -> expectString(n, nodeKey + " entry"))
+                       .collect(toList());
+        }
+
+        throw new IllegalArgumentException(nodeKey + " was not an array or string");
+    }
+
+    private String getString(Map<String, JsonNode> object, String key) {
+        JsonNode node = object.get(key);
+        if (node == null) {
+            return null;
+        }
+
+        return expectString(node, key);
+    }
+
+    private String expectString(JsonNode node, String name) {
+        Validate.isTrue(node.isString(), "%s was not a string");
+        return node.asString();
```

Comment:
name missing as argument:
`Validate.isTrue(node.isString(), "%s was not a string", name);`

Reply:
Good catch!